Add single version migrations support to DashboardMigrator

ivanortegaalba opened the pull request:
Add support for single version migrations in DashboardMigrator, allowing migration to specific target versions instead of always migrating to the latest schema version.

## Problem

DashboardMigrator currently only supports migrating to the latest schema version (v41), making it difficult to test individual migration steps and debug migration issues for specific versions.

## Solution

- Add targetVersion parameter to DashboardMigrator.updateSchema() and DashboardModel constructor
- Create separate output directories (single_version/ vs latest_version/) to prevent test confusion
- Fix migration logic to respect target version (prevents running migrations beyond target)
- Add comprehensive test suite comparing frontend vs backend single version migrations
- Remove old v14 broken dashboard files that violated naming convention

## How to test

1. Run backend single version tests:
   go test -v ./apps/dashboard/pkg/migration/ -run TestMigrateSingleVersion

2. Run frontend single version tests:
   yarn test public/app/features/dashboard/state/DashboardMigratorSingleVersion.test.ts --no-watch

3. Verify that both test suites pass and generate appropriate output files in their respective directories.

ivanortegaalba committed:
Add single version migrations support to DashboardMigrator

- Add targetVersion parameter to DashboardMigrator.updateSchema() and DashboardModel constructor
- Create separate output directories (single_version/ vs latest_version/) to prevent test confusion
- Fix migration logic to respect target version (prevents running migrations beyond target)
- Add comprehensive test suite comparing frontend vs backend single version migrations
- Remove old v14 broken dashboard files that violated naming convention

ivanortegaalba committed:
Fix DashboardMigratorToBackend.test.ts to use correct output directory structure

- Update test to read input files from inputDir instead of outputDir
- Fix backend output filename construction to match new naming convention
- Ensure test reads from latest_version subdirectory for backend outputs

ivanortegaalba committed:
Only first 10 versions

dprokop commented on apps/dashboard/pkg/migration/migrate_test.go:
```diff
@@ -66,11 +84,75 @@ func TestMigrate(t *testing.T) {
+			t.Fatalf("input filename must use v{N}.{name}.json format, got: %s", f.Name())
+		}
+
+		// Extract version from filename (e.g., v16.grid_layout_upgrade.json -> 16)
+		// This represents the tested version that the file should be migrated TO
+		versionStr := strings.TrimPrefix(f.Name(), "v")
+		dotIndex := strings.Index(versionStr, ".")
+		if dotIndex == -1 {
+			t.Fatalf("input filename must use v{N}.{name}.json format, got: %s", f.Name())
+		}
```
This is incorrect, the migration shoult be to v17 in this example case.

dprokop replied:
Ah, actually it is, because the chema version in the `vX` file is `v(X-1)`. This is confusing :D

dprokop commented on public/app/features/dashboard/state/DashboardMigratorSingleVersion.test.ts:
```diff
@@ -0,0 +1,270 @@
+ */
+
+// Set up the same datasources as backend test provider to ensure consistency
+const dataSources = {
```
Let's extract this to some sort of util, it's used in two tests now.

dprokop commented on public/app/features/dashboard/state/DashboardMigratorSingleVersion.test.ts:
```diff
@@ -0,0 +1,270 @@
+        which means that the actual migration logic is not run.
+        We need to manually run the pluginLoaded logic to ensure the panels are migrated correctly.
+        */
+        if (jsonInput.schemaVersion <= 27) {
```
let's extract this to an util too.

dprokop commented on public/app/features/dashboard/state/DashboardMigratorSingleVersion.test.ts:
```diff
@@ -0,0 +1,270 @@
+    });
+});
+
+function cleanDashboardModel(dashboard: DashboardModel) {
```
And this :)

ivanortegaalba committed:
Add tests for specific target version migration functionality

- Add tests to DashboardMigrator.test.ts for migrating to specific target versions
- Add tests to DashboardModel.test.ts for targetSchemaVersion option
- Tests verify target version migration works correctly
- Tests ensure no extra migrations when target equals current
- All existing tests remain functional and green

ivanortegaalba committed:
Refactor migration test files to reuse common code

- Extract common utilities to migrationTestUtils.ts
- Refactor DashboardMigratorSingleVersion.test.ts to use shared utilities
- Refactor DashboardMigratorToBackend.test.ts to use shared utilities
- Reduce code duplication and improve maintainability
- All existing tests continue to pass

ivanortegaalba committed:
Refactor migration tests and fix DashboardMigrator schema version bug

- Extract common utilities to migrationTestUtils.ts for code reuse
- Refactor DashboardMigratorSingleVersion.test.ts to use shared utilities
- Refactor DashboardMigratorToBackend.test.ts to use shared utilities
- Fix critical bug where schema version was not updated for dashboard-level migrations
- Move schema version update before panel upgrades check to ensure proper updates
- Only update schema version when migrating to higher version to prevent inconsistent state
- Update test expectations to reflect correct schema version behavior
- All 270 tests now pass with correct implementation

ivanortegaalba committed:
Remove unnecessary option
